File: app/src/main/java/com/thatcakeid/zrytezene/ui/login/LoginFragment.java
package com.thatcakeid.zrytezene.ui.login;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.thatcakeid.zrytezene.R;

public class LoginFragment extends Fragment {
    private View view;
    private TextInputEditText tie1, tie2;
    private MaterialButton button_continue;
    private FirebaseAuth auth;

    public static LoginFragment newInstance() {
        return new LoginFragment();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.login_fragment, container, false);
        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        tie1 = view.findViewById(R.id.tie1);
        tie2 = view.findViewById(R.id.tie2);
        button_continue = view.findViewById(R.id.button_continue);

        button_continue.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (tie1.getText().toString().trim().length() == 0)
                    Snackbar.make(view, "Email field can't be empty!", Snackbar.LENGTH_LONG);
                else {
                    if (tie2.getText().toString().length() == 0)
                        Snackbar.make(view, "Password field can't be empty!", Snackbar.LENGTH_LONG);
                    else {
                        auth.createUserWithEmailAndPassword(tie1.getText().toString().trim(),
                                tie2.getText().toString()).addOnSuccessListener(new OnSuccessListener<AuthResult>() {
                            @Override
                            public void onSuccess(AuthResult authResult) {
                                auth.getCurrentUser().sendEmailVerification().addOnSuccessListener(new OnSuccessListener<Void>() {
                                    @Override
                                    public void onSuccess(Void aVoid) {
                                        Snackbar.make(view, "A verification link has been sent to your email. Please check your inbox or spam box.", Snackbar.LENGTH_LONG);
                                    }
                                }).addOnFailureListener(new OnFailureListener() {
                                    @Override
                                    public void onFailure(@NonNull Exception e) {
                                        Snackbar.make(view, e.getMessage(), Snackbar.LENGTH_LONG);
                                    }
                                });
                            }
                        }).addOnFailureListener(new OnFailureListener() {
                            @Override
                            public void onFailure(@NonNull Exception e) {
                                Snackbar.make(view, e.getMessage(), Snackbar.LENGTH_LONG);
                            }
                        });
                    }
                }
            }
        });
    }

}

File: app/src/main/java/com/thatcakeid/zrytezene/ui/startup/LoginFragment.kt
package com.thatcakeid.zrytezene.ui.startup

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.findNavController
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import com.thatcakeid.zrytezene.R
import com.thatcakeid.zrytezene.databinding.FragmentLoginBinding
import com.zhuinden.fragmentviewbindingdelegatekt.viewBinding

// TODO: 10/5/21 move stuff to a viewmodel
class LoginFragment : Fragment(R.layout.fragment_login) {
    private val binding: FragmentLoginBinding by viewBinding(FragmentLoginBinding::bind)
    private val auth by lazy { FirebaseAuth.getInstance() }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.loginEmailTie.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun afterTextChanged(s: Editable) {}

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (binding.loginEmailTie.text.toString().trim()
                        .matches("[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+".toRegex())) {
                    binding.loginEmailTil.error = null
                } else {
                    binding.loginEmailTil.error = "Invalid email!"
                }
            }
        })

        binding.buttonContinue.setOnClickListener {
            val email = binding.loginEmailTie.text.toString().trim()
            val password = binding.loginPasswTie.text.toString()

            if (!email.matches("[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+".toRegex())) {
                Toast.makeText(this@LoginFragment.requireContext(), "Invalid email!", Toast.LENGTH_LONG).show()

                return@setOnClickListener
            }

            if (password.isEmpty()) {
                Snackbar.make(binding.root, "Password can't be empty!", Snackbar.LENGTH_LONG).show()

                return@setOnClickListener
            }

            auth.signInWithEmailAndPassword(email, password).addOnSuccessListener {
                if (auth.currentUser!!.isEmailVerified) {
                    Toast.makeText(this@LoginFragment.requireContext(), "Logged in", Toast.LENGTH_LONG)
                        .show()

                    findNavController()
                        .popBackStack()

                } else {
                    val alertDialog = AlertDialog.Builder(this@LoginFragment.requireContext())

                    alertDialog.setTitle("Unverified Email")
                    alertDialog.setMessage("Your email isn't verified. Do you want to re-send a new verification link to your email?")
                    alertDialog.setPositiveButton("Yes") { _, _ ->
                        auth.currentUser!!.sendEmailVerification().addOnSuccessListener {
                                Snackbar.make(
                                    binding.root,
                                    "A verification link has been sent to your email. Please check your inbox or spam box.",
                                    Snackbar.LENGTH_LONG
                                ).show()
                            }
                            .addOnFailureListener { e ->
                                Snackbar.make(
                                    binding.root,
                                    e.message!!,
                                    Snackbar.LENGTH_LONG
                                ).show()
                            }
                    }

                    alertDialog.setNegativeButton("No") { _, _ -> auth.signOut() }
                    alertDialog.setCancelable(false)
                    alertDialog.create().show()
                }
            }.addOnFailureListener { e ->
                Snackbar.make(
                    binding.root,
                    e.message!!,
                    Snackbar.LENGTH_LONG
                ).show()
            }
        }

        binding.forgotPasswordText.setOnClickListener {
            val email = binding.loginEmailTie.text.toString().trim()

            if (email.matches("[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+".toRegex())) {
                Toast.makeText(this@LoginFragment.requireContext(), "Please enter a valid email!", Toast.LENGTH_LONG)
                    .show()
                return@setOnClickListener
            }

            auth.sendPasswordResetEmail(email)
            Snackbar.make(
                binding.root,
                "A verification link has been sent to your email. Please check your inbox or spam box.",
                Snackbar.LENGTH_LONG
            ).show()

        }

        binding.registerText.setOnClickListener {
            findNavController()
                .navigate(
                    LoginFragmentDirections.actionLoginFragmentToRegisterFragment(
                        binding.loginEmailTie.text.toString().trim()
                    )
                )
        }
    }
}